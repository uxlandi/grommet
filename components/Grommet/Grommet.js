"use strict";

exports.__esModule = true;
exports.Grommet = void 0;

var _react = _interopRequireWildcard(require("react"));

var _contexts = require("grommet-icons/contexts");

var _recompose = require("recompose");

var _contexts2 = require("../../contexts");

var _base = require("../../themes/base");

var _utils = require("../../utils");

var _hocs = require("../hocs");

var _StyledGrommet = require("./StyledGrommet");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

// grommet-icons isn't aware of the grommet dark background context.
// Here, we reduce the grommet theme colors to the correct flat color
// namespace for grommet-icons.
var reduceIconTheme = function reduceIconTheme(iconTheme, dark) {
  var result = _extends({}, iconTheme, {
    colors: _extends({}, iconTheme.colors)
  });

  Object.keys(result.colors).forEach(function (key) {
    if (typeof result.colors[key] === 'object') {
      result.colors[key] = (0, _utils.normalizeColor)(result.colors[key][dark ? 'dark' : 'light'], {
        dark: dark,
        global: {
          colors: result.colors
        }
      });
    } else {
      result.colors[key] = (0, _utils.normalizeColor)(result.colors[key], {
        dark: dark,
        global: {
          colors: result.colors
        }
      });
    }
  });
  return result;
};

var Grommet =
/*#__PURE__*/
function (_Component) {
  _inheritsLoose(Grommet, _Component);

  function Grommet() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "state", {});

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onResize", function () {
      var _this$state = _this.state,
          theme = _this$state.theme,
          responsive = _this$state.responsive;
      var breakpoint = (0, _utils.getBreakpoint)(window.innerWidth, theme);

      if (breakpoint !== responsive) {
        _this.setState({
          responsive: breakpoint
        });
      }
    });

    return _this;
  }

  Grommet.getDerivedStateFromProps = function getDerivedStateFromProps(nextProps, prevState) {
    var iconTheme = nextProps.iconTheme,
        theme = nextProps.theme;
    var stateTheme = prevState.theme,
        themeProp = prevState.themeProp,
        iconThemeProp = prevState.iconThemeProp;
    var nextTheme;

    if (theme && (theme !== themeProp || iconTheme !== iconThemeProp)) {
      // in case the supplied theme has global.colors but not icon.colors,
      // pre-merge the current base icon colors with the new theme colors.
      var iconColoredTheme = theme;

      if (!theme.icon || !theme.icon.colors) {
        iconColoredTheme = _extends({}, theme);
        iconColoredTheme.icon = _extends({}, theme.icon || {});
        iconColoredTheme.icon.colors = (0, _utils.deepMerge)(_base.base.icon.colors, (theme.global || {}).colors);
      }

      nextTheme = (0, _utils.deepMerge)(_base.base, iconColoredTheme);
    } else if (!theme && (themeProp || !stateTheme)) {
      nextTheme = _base.base;
    }

    if (nextTheme) {
      var _ref = nextTheme.global || _base.base.global,
          colors = _ref.colors;

      var color = colors.background;
      var dark = color ? (0, _utils.colorIsDark)(color) : false;
      var lightIconTheme = (0, _utils.deepMerge)(iconTheme, nextTheme.icon);
      var iconThemes = {
        dark: reduceIconTheme((0, _utils.deepMerge)(lightIconTheme, {
          color: colors.text.dark
        }), true),
        light: reduceIconTheme(lightIconTheme, false)
      };
      return {
        theme: _extends({}, nextTheme, {
          dark: dark,
          icon: dark ? iconThemes.dark : iconThemes.light,
          iconThemes: iconThemes
        }),
        themeProp: theme,
        iconThemeProp: iconTheme
      };
    }

    return null;
  };

  var _proto = Grommet.prototype;

  _proto.componentDidMount = function componentDidMount() {
    window.addEventListener('resize', this.onResize);
    this.onResize();
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  };

  _proto.render = function render() {
    var _this$props = this.props,
        children = _this$props.children,
        rest = _objectWithoutPropertiesLoose(_this$props, ["children"]);

    delete rest.theme;
    var _this$state2 = this.state,
        responsive = _this$state2.responsive,
        theme = _this$state2.theme;
    return _react.default.createElement(_contexts2.ThemeContext.Provider, {
      value: theme
    }, _react.default.createElement(_contexts.ThemeContext.Provider, {
      value: theme.icon
    }, _react.default.createElement(_contexts2.ResponsiveContext.Provider, {
      value: responsive
    }, _react.default.createElement(_StyledGrommet.StyledGrommet, _extends({}, rest, {
      theme: theme
    }), children))));
  };

  return Grommet;
}(_react.Component);

var GrommetDoc;

if (process.env.NODE_ENV !== 'production') {
  GrommetDoc = require('./doc').doc(Grommet); // eslint-disable-line global-require
}

var GrommetWrapper = (0, _recompose.compose)(_hocs.withIconTheme)(GrommetDoc || Grommet);
exports.Grommet = GrommetWrapper;