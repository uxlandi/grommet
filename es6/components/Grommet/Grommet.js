function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { Component } from 'react';
import { ThemeContext as IconThemeContext } from "grommet-icons/es6/contexts/ThemeContext";
import { compose } from 'recompose';
import { ResponsiveContext, ThemeContext } from '../../contexts';
import { base as baseTheme } from '../../themes/base';
import { colorIsDark, deepMerge, getBreakpoint, normalizeColor } from '../../utils';
import { withIconTheme } from '../hocs';
import { StyledGrommet } from './StyledGrommet'; // grommet-icons isn't aware of the grommet dark background context.
// Here, we reduce the grommet theme colors to the correct flat color
// namespace for grommet-icons.

var reduceIconTheme = function reduceIconTheme(iconTheme, dark) {
  var result = _extends({}, iconTheme, {
    colors: _extends({}, iconTheme.colors)
  });

  Object.keys(result.colors).forEach(function (key) {
    if (typeof result.colors[key] === 'object') {
      result.colors[key] = normalizeColor(result.colors[key][dark ? 'dark' : 'light'], {
        dark: dark,
        global: {
          colors: result.colors
        }
      });
    } else {
      result.colors[key] = normalizeColor(result.colors[key], {
        dark: dark,
        global: {
          colors: result.colors
        }
      });
    }
  });
  return result;
};

var Grommet =
/*#__PURE__*/
function (_Component) {
  _inheritsLoose(Grommet, _Component);

  function Grommet() {
    var _this;

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _Component.call.apply(_Component, [this].concat(args)) || this;

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "state", {});

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "onResize", function () {
      var _this$state = _this.state,
          theme = _this$state.theme,
          responsive = _this$state.responsive;
      var breakpoint = getBreakpoint(window.innerWidth, theme);

      if (breakpoint !== responsive) {
        _this.setState({
          responsive: breakpoint
        });
      }
    });

    return _this;
  }

  Grommet.getDerivedStateFromProps = function getDerivedStateFromProps(nextProps, prevState) {
    var iconTheme = nextProps.iconTheme,
        theme = nextProps.theme;
    var stateTheme = prevState.theme,
        themeProp = prevState.themeProp,
        iconThemeProp = prevState.iconThemeProp;
    var nextTheme;

    if (theme && (theme !== themeProp || iconTheme !== iconThemeProp)) {
      // in case the supplied theme has global.colors but not icon.colors,
      // pre-merge the current base icon colors with the new theme colors.
      var iconColoredTheme = theme;

      if (!theme.icon || !theme.icon.colors) {
        iconColoredTheme = _extends({}, theme);
        iconColoredTheme.icon = _extends({}, theme.icon || {});
        iconColoredTheme.icon.colors = deepMerge(baseTheme.icon.colors, (theme.global || {}).colors);
      }

      nextTheme = deepMerge(baseTheme, iconColoredTheme);
    } else if (!theme && (themeProp || !stateTheme)) {
      nextTheme = baseTheme;
    }

    if (nextTheme) {
      var _ref = nextTheme.global || baseTheme.global,
          colors = _ref.colors;

      var color = colors.background;
      var dark = color ? colorIsDark(color) : false;
      var lightIconTheme = deepMerge(iconTheme, nextTheme.icon);
      var iconThemes = {
        dark: reduceIconTheme(deepMerge(lightIconTheme, {
          color: colors.text.dark
        }), true),
        light: reduceIconTheme(lightIconTheme, false)
      };
      return {
        theme: _extends({}, nextTheme, {
          dark: dark,
          icon: dark ? iconThemes.dark : iconThemes.light,
          iconThemes: iconThemes
        }),
        themeProp: theme,
        iconThemeProp: iconTheme
      };
    }

    return null;
  };

  var _proto = Grommet.prototype;

  _proto.componentDidMount = function componentDidMount() {
    window.addEventListener('resize', this.onResize);
    this.onResize();
  };

  _proto.componentWillUnmount = function componentWillUnmount() {
    window.removeEventListener('resize', this.onResize);
  };

  _proto.render = function render() {
    var _this$props = this.props,
        children = _this$props.children,
        rest = _objectWithoutPropertiesLoose(_this$props, ["children"]);

    delete rest.theme;
    var _this$state2 = this.state,
        responsive = _this$state2.responsive,
        theme = _this$state2.theme;
    return React.createElement(ThemeContext.Provider, {
      value: theme
    }, React.createElement(IconThemeContext.Provider, {
      value: theme.icon
    }, React.createElement(ResponsiveContext.Provider, {
      value: responsive
    }, React.createElement(StyledGrommet, _extends({}, rest, {
      theme: theme
    }), children))));
  };

  return Grommet;
}(Component);

var GrommetDoc;

if (process.env.NODE_ENV !== 'production') {
  GrommetDoc = require('./doc').doc(Grommet); // eslint-disable-line global-require
}

var GrommetWrapper = compose(withIconTheme)(GrommetDoc || Grommet);
export { GrommetWrapper as Grommet };