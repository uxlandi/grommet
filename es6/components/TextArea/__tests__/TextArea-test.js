import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';
import { Grommet } from '../../Grommet';
import { TextArea } from '..';
jest.mock('react-dom');
describe('TextArea', function () {
  test('basic', function () {
    var component = renderer.create(React.createElement(Grommet, null, React.createElement(TextArea, {
      id: "item",
      name: "item"
    })));
    var tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('placeholder', function () {
    var component = renderer.create(React.createElement(Grommet, null, React.createElement(TextArea, {
      id: "item",
      name: "item",
      placeholder: "placeholder"
    })));
    var tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('plain', function () {
    var component = renderer.create(React.createElement(Grommet, null, React.createElement(TextArea, {
      id: "item",
      name: "item",
      plain: true
    })));
    var tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('focusIndicator', function () {
    var component = renderer.create(React.createElement(Grommet, null, React.createElement(TextArea, {
      id: "item",
      name: "item",
      focusIndicator: true
    })));
    var tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('fill', function () {
    var component = renderer.create(React.createElement(Grommet, null, React.createElement(TextArea, {
      id: "item",
      name: "item",
      fill: true
    })));
    var tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});