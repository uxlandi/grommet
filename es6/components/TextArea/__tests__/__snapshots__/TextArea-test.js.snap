// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea basic 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  border: none;
  -webkit-appearance: none;
  padding: 11px;
  outline: none;
  background: transparent;
  color: inherit;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0,0,0,0.33);
  border-radius: 4px;
  width: 100%;
}

.c1::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c1::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c1::-moz-placeholder {
  color: #AAAAAA;
}

.c1:-ms-input-placeholder {
  color: #AAAAAA;
}

.c1::-moz-focus-inner {
  border: none;
  outline: none;
}

<div
  className="c0"
>
  <textarea
    className="c1"
    id="item"
    name="item"
    onBlur={[Function]}
    onFocus={[Function]}
  />
</div>
`;

exports[`TextArea fill 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  border: none;
  -webkit-appearance: none;
  padding: 11px;
  outline: none;
  background: transparent;
  color: inherit;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0,0,0,0.33);
  border-radius: 4px;
  width: 100%;
  height: 100%;
}

.c1::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c1::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c1::-moz-placeholder {
  color: #AAAAAA;
}

.c1:-ms-input-placeholder {
  color: #AAAAAA;
}

.c1::-moz-focus-inner {
  border: none;
  outline: none;
}

<div
  className="c0"
>
  <textarea
    className="c1"
    id="item"
    name="item"
    onBlur={[Function]}
    onFocus={[Function]}
  />
</div>
`;

exports[`TextArea focusIndicator 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  border: none;
  -webkit-appearance: none;
  padding: 11px;
  outline: none;
  background: transparent;
  color: inherit;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0,0,0,0.33);
  border-radius: 4px;
  width: 100%;
}

.c1::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c1::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c1::-moz-placeholder {
  color: #AAAAAA;
}

.c1:-ms-input-placeholder {
  color: #AAAAAA;
}

.c1::-moz-focus-inner {
  border: none;
  outline: none;
}

<div
  className="c0"
>
  <textarea
    className="c1"
    id="item"
    name="item"
    onBlur={[Function]}
    onFocus={[Function]}
  />
</div>
`;

exports[`TextArea placeholder 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  border: none;
  -webkit-appearance: none;
  padding: 11px;
  outline: none;
  background: transparent;
  color: inherit;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0,0,0,0.33);
  border-radius: 4px;
  width: 100%;
}

.c1::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c1::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c1::-moz-placeholder {
  color: #AAAAAA;
}

.c1:-ms-input-placeholder {
  color: #AAAAAA;
}

.c1::-moz-focus-inner {
  border: none;
  outline: none;
}

<div
  className="c0"
>
  <textarea
    className="c1"
    id="item"
    name="item"
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder="placeholder"
  />
</div>
`;

exports[`TextArea plain 1`] = `
.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

.c1 {
  box-sizing: border-box;
  font-size: inherit;
  border: none;
  -webkit-appearance: none;
  padding: 11px;
  outline: none;
  background: transparent;
  color: inherit;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0,0,0,0.33);
  border-radius: 4px;
  width: 100%;
  border: none;
  width: 100%;
  -webkit-appearance: none;
}

.c1::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c1::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c1::-moz-placeholder {
  color: #AAAAAA;
}

.c1:-ms-input-placeholder {
  color: #AAAAAA;
}

.c1::-moz-focus-inner {
  border: none;
  outline: none;
}

<div
  className="c0"
>
  <textarea
    className="c1"
    id="item"
    name="item"
    onBlur={[Function]}
    onFocus={[Function]}
  />
</div>
`;
